src/coreclr/jit/flowgraph.cpp

```diff
@@ -1270,6 +1270,44 @@ bool Compiler::fgCastNeeded(GenTree* tree, var_types toType)
     return true;
 }
 
+//-------------------------------------------------------------------------------------
+// fgCastRequiresHelper: Check whether a given cast must be converted to a helper call.
+//
+// Arguments:
+//    fromType - The source type of the cast.
+//    toType   - The target type of the cast.
+//    overflow - True if the cast has the GTF_OVERFLOW flag set.
+//
+// Return Value:
+//    True if the cast requires a helper call, otherwise false.
+//
+bool Compiler::fgCastRequiresHelper(var_types fromType, var_types toType, bool overflow /* false */)
+{
+    if (varTypeIsFloating(fromType))
+    {
+        return (varTypeIsIntegral(toType) && overflow)
+#if defined(TARGET_X86)
+               || (varTypeIsLong(toType) && !compOpportunisticallyDependsOn(InstructionSet_AVX512))
+#elif !defined(TARGET_64BIT)
+               || varTypeIsLong(toType)
+#endif
```

Comment:
nit: I find the current logic a lot harder to follow, due to the single return statement spanning across multiple ifdefs. I think something like the following, although longer, makes it much clearer as to what is happening:
```cpp
if (varTypeIsFloating(fromType))
{
    if (overflow)
    {
        assert(varTypeIsIntegral(toType));
        return true;
    }

#if !defined(TARGET_64BIT)
    if (varTypeIsLong)
    {
#if defined(TARGET_X86)
        if (compOpportunisticallyDependsOn(InstructionSet_AVX512))
        {
            return false;
        }
#endif

        return true;
    }
#endif

    return false;
}
```